cmd/namespace-lock.go

```diff
@@ -96,16 +93,21 @@ func (n *nsLockMap) lock(ctx context.Context, volume string, path string, lockSo
 		locked = nsLk.GetLock(opsID, lockSource, timeout)
 	}
 
+	n.lockMapMutex.Lock()
 	if !locked { // We failed to get the lock
-
 		// Decrement ref count since we failed to get the lock
-		if atomic.AddUint32(&nsLk.ref, ^uint32(0)) == 0 {
+		if atomic.AddUint32(&n.lockMap[resource].ref, ^uint32(0)) == 0 {
 			// Remove from the map if there are no more references.
-			n.lockMapMutex.Lock()
 			delete(n.lockMap, resource)
-			n.lockMapMutex.Unlock()
+		}
+	} else {
+		// It is possible a parallel unlock has reduced the refCount.
```

Comment:
This doesn't make sense to me. Since we have incremented by 1, how could an unlock ever make it 0 since that would indicate that it never had the lock in the first place.

You can keep the check, but it should be a panic since AFAICT it would be a leaking use of the locks.

Reply:
It can because nsLk is actually a shared map value @klauspost it is possible for `unlock()` function to be called and unlock a locker which was active before we got here.

Reply:
so we need to make sure that we always increment before we return from this function upon success